src/pip/_internal/utils/misc.py

```diff
@@ -852,6 +853,27 @@ def enum(*sequential, **named):
     return type('Enum', (), enums)
 
 
+def redact_auth_from_url(url):
+    # Return a copy of url with 'username:password@' redacted by
+    # substituting the credentials with '<redacted>' in case they
+    # were present in the url.
```

Comment:
I think it would be better if only the password is replaced with a substitution string (e.g. `<redacted>` or `****`). It helps to have the username present, e.g. for troubleshooting purposes.